resource_elasticsearch_odfe_policy.go

```diff
@@ -0,0 +1,227 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/url"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/structure"
+	"github.com/olivere/elastic/uritemplates"
+
+	elastic7 "github.com/olivere/elastic/v7"
+)
+
+func resourceElasticsearchOdfePolicy() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceElasticsearchOdfePolicyCreate,
+		Read:   resourceElasticsearchOdfePolicyRead,
+		Update: resourceElasticsearchOdfePolicyCreate,
+		Delete: resourceElasticsearchOdfePolicyDelete,
+		Schema: map[string]*schema.Schema{
+			"policy_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"body": &schema.Schema{
+				Type:             schema.TypeString,
+				Required:         true,
+				DiffSuppressFunc: diffSuppressPolicy,
+				StateFunc: func(v interface{}) string {
+					json, _ := structure.NormalizeJsonString(v)
+					return json
+				},
+			},
+			"primary_term": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Computed: true,
+			},
+			"seq_no": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Computed: true,
+			},
+		},
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+	}
+}
+
+func resourceElasticsearchOdfePolicyCreate(d *schema.ResourceData, m interface{}) error {
+	if _, err := resourceElasticsearchPutOdfePolicy(d, m); err != nil {
+		return err
+	}
+
+	return resourceElasticsearchOdfePolicyRead(d, m)
+}
+
+func resourceElasticsearchOdfePolicyRead(d *schema.ResourceData, m interface{}) error {
+	policyID := d.Get("policy_id").(string)
+	policyResponse, err := resourceElasticsearchGetOdfePolicy(policyID, m)
+
+	if err != nil {
+		if elastic7.IsNotFound(err) {
+			log.Printf("[WARN] OdfePolicy (%s) not found, removing from state", policyID)
+			d.SetId("")
+			return nil
+		}
+	}
+
+	bodyString, err := json.Marshal(policyResponse.Policy)
+	// Need encapsulation as the reponse from the GET is different than the one in the PUT
+	bodyStringNormalized, _ := structure.NormalizeJsonString(fmt.Sprintf("{\"policy\": %+s}", string(bodyString)))
+
+	if err != nil {
+		return err
+	}
+	d.SetId(policyResponse.PolicyID)
+	d.Set("body", bodyStringNormalized)
```

Comment:
body isn't a computed value, should it really be set ?

Reply:
yes, like so we update the state to reflect the real state of the object, just in case of manual change .